### ISSUE TITLE
Threaded indexing

### ISSUE THREAD

**Issue by zevv** [zevv/duc#12]
Investigate methods for multi-threaded indexing

**Comment by zevv** [comment]
This is one of the major missing features in Duc. A lot of larger setup would benefit significantly by doing the indexing in parallel. Various options we can consider:

a) Make the indexer process truly multi threaded. This will require a fancy algorithm which is able to split up the tree into subtrees, spawn worker threads for each subtree and properly add up the running totals of the common trunks of the trees.

b) Run separate indexers in parallel on different trees, and provide the tools to merge the resulting indexes into a single index

c) Run separate indexers in parallel on different trees, and update the displaying tools to be able to query more then one database, showing the result as if the data comes from a single database.

I feel that option a) is the holy grail, as this will hide all the implementation details to the user and will 'just work' in most cases.

Any comments welcome.

**Comment by digitalman2112** [comment]
Thoughts:

Agreed that #1 is optimal and hides the details. 

I have a number of use cases that would benefit from the ability to treat aggregate multiple databases as a single view (#3) should this become possible - but that could be decided independently from how you solve for indexing.

**Comment by pol** [comment]
Out of curiosity, is it plausible to fake this by running multiple duc index jobs simultaneously writing to the same database? Would that mess with the database?  Basically, would it be better to find some other tool that can split arbitrary filesystem trees into N fairly-equal subtrees and start duc index jobs in each root?  This method would also make extremely quick work of a massive filesystem on a compute cluster (which is what I'm using duc for).

**Comment by zevv** [comment]
* On 2015-04-24 06:03:15 +0200, pol llovet wrote:
 
> Out of curiosity, is it plausible to fake this by running multiple duc
> index jobs simultaneously writing to the same database? Would that
> mess with the database?  

That would mess up the database indeed. There's a few complicating
factors here. First one is that the underlying tokyocabinet has a global
reader/writer lock. This means there's only one process allowed to
write, and during the write all readers are blocked. The other problem
is that the order the directory tree is traversed does not allow for
easy parallization. Consider this directory tree:

/etc
 `+- X11
  |  `+- fonts
  |   |- server
  |   `- xsm
  `- pam.d
      `- security

To traverse a (directory tree), you start at the root and recursively descend
into the subdirectores. The total file size of a directory (for example, /etc)
can only be calculated when you know the total size of all underying
directories. If we would now simply parallelize the indexing, one process
indexing /etc/X11 and another indexing /etc/pam.d. Only when both are complete
and done, you can calculate the total size of /etc.

The 'threaded-indexing' branch in the git repo takes a threaded/recursive
approach: it has a work queue in which jobs are put to calculate the size of a
directory. When a new subdirectory is found by a job, it creates a new job to
index that subdirectory. When the size of a subdirectory is known, the results
are accumulated in the job of its parent directory, and as soon as all parts of
a job is complete, its record is written to the database.

This approach works reasonably well, but I'm a bit put off by the resulting
complexity. It makes it hard to reason about correctness and is much more hairy
then the simple recursive approach.

Currently I'm considering a different method: we simply spawn multiple jobs for
multiple subdirectories, and all these subdirectories are calculated
separately, without any accumulation on their common ancestor. Instead the
accumulation is done at query-time by simply finding all entries in the
database that apply for a given directory.  This would very much simplify the
indexing process, at the cost of a bit more work when querying.

I've done some tests with multiple threads, and it seems that there's virutally
nothing to win to run multiple threads to index a plain file system on a single
disk/spindle. Things change however for lvm and RAID systems, and change a lot
when there's file systems which are distributed over lots of different nodes in
a network. 

> Basically, would it be better to find some other tool that can split
> arbitrary filesystem trees into N fairly-equal subtrees and start duc
> index jobs in each root?  This method would also make extremely quick
> work of a massive filesystem on a compute cluster (which is what I'm
> using duc for).

That kind of systems is where Duc should shine, so I still hope to find
a proper solution for this. Stay tuned!
 

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by pol** [comment]
As far as recursive thread spawning, wouldn't it make more sense to do it bottom-up instead of top-down?  For example, let's say you are recursing from left to right (probably not the most efficient), you hit the bottom of the left side of the tree, and start a job for that directory (trivial), then pop up to the parent dir, dive again and if there is threadpool room, start an index of that leaf. 

I don't really know anything about tokyocabinet, but perhaps the dbd's created in those initial dirs can bubble up and act as seeds for the above index, making it so that the parent index only needs to trivially descend (check). Can TC dbd's be merged easily?

If you hit MAXTHREADS, then you pause until a slot opens up.

It seems like trying to start a thread for subdirs as you make your way down would be more prone to unbalancing the job.  I could be wrong and it may be identical algorithmically in the long run, but my gut likes it better.

I started looking up some algorithms for splitting arbitrary DAGs into roughly equal sub DAGs, and all of them seem to involve walking the entire graph, which sort of defeats the purpose for us.

**Comment by zevv** [comment]
* On 2015-04-24 19:36:28 +0200, pol llovet wrote:

> As far as recursive thread spawning, wouldn't it make more sense to do
> it bottom-up instead of top-down? For example, let's say you are
> recursing from left to right (probably not the most efficient), you
> hit the bottom of the left side of the tree, and start a job for that
> directory (trivial), then pop up to the parent dir, dive again and if
> there is threadpool room, start an index of that leaf.

Hm yes, that makes sense, although I'm not sure if that would make the
overall implemenation any more simple then my initial try. Also gut
feelings here, I usually need to get down to making prototypes before I
can oversee all the implications of these things.
 
> I started looking up some algorithms for splitting arbitrary DAGs into
> roughly equal sub DAGs, and all of them seem to involve walking the
> entire graph, which sort of defeats the purpose for us. 

That's exactly my problem, because walking the tree is the expensive
operation here.

My original idea was simple: just have N threads for each unique file
system, and as soon a new file system is reached (dev_t should change),
spawn N extra threads that work only directories with that dev_t.

N can be small, probably even just 1, since I have not measured a
practical speedup on my test systems when scanning a file system with
more then 1 thread.

There's one more very important complicating factor here, and that is
that I'm simply too stupid to work with threads :(

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by pol** [comment]
How hard is it to merge N tokyocabinet dbs?  I think the "external tool that calls duc" might be a more fruitful direction.

You'd think that the filesystem would have some kind of metadata somewhere that could give you an idea of the "middle" of the fs. 'course that will be extremely filesystem dependent. It's nice that duc doesn't seem to care... or does it?

I just tried to run duc on a gluster filesystem, and it did not seem to be happy.  Here's the screencap:

```
[root@gstore2 clients]# time duc index -p earthscience
Error statting chrome.gif: No data availabletories
^C-----#-] Indexed 0b in 0 files and 0 directories

real    213m5.035s
user    0m2.934s
sys     0m17.431s
[root@gstore2 clients]# cat /etc/ducrc
database /mnt/gluster/durable/rcg/duc/gstore.ducdb
[root@gstore2 clients]# ls -lah /mnt/gluster/durable/rcg/duc/gstore.ducdb
-rw-r--r--. 1 root root 133K Apr 24 11:41 /mnt/gluster/durable/rcg/duc/gstore.ducdb
```

I would have expected more to be pouring into that ducdb file. Any tips?

**Comment by zevv** [comment]
* On 2015-04-24 23:17:06 +0200, pol llovet wrote:
 
> How hard is it to merge N tokyocabinet dbs?  I think the "external
> tool that calls duc" might be a more fruitful direction.

Yeah, that might prove more robust. I'll have to think about the
merging, it's not so much merging-of-tokyocabinet-db's, but more just
the merging-of-trees.

> You'd think that the filesystem would have some kind of metadata
> somewhere that could give you an idea of the "middle" of the fs.
> 'course that will be extremely filesystem dependent. It's nice that
> duc doesn't seem to care... or does it?

It would be nice if it could do The Right Thing in 90% of the cases, and
for this it would probably need to know a bit about what file systems it
is running on. But overdoing this kind of heuristics is a of course
dangerous.

> I just tried to run duc on a gluster filesystem, and it did not seem
> to be happy.  Here's the screencap:
> 
> ``` [root@gstore2 clients]# time duc index -p earthscience Error
> statting chrome.gif: No data availabletories ^C-----#-] Indexed 0b in
> 0 files and 0 directories

Don't be fooled by the progress meter showing only 0's, this was
recently fixed. Running 'duc info' after indexing should show you
what is really in the database.

> I would have expected more to be pouring into that ducdb file. Any
> tips?

'duc info', 'duc index --verbose', or 'duc index --debug' should tell
you much more.

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by pol** [comment]
This is probably the wrong place for this, but something is awry. Should I create a new issue?

so, it's interesting.  It's clearly doing something (based on the debug info), but I don't think anything is making it into the DB. 

```
[root@gstore2 clients]# duc info
[root@gstore2 clients]# duc info earthscience
[root@gstore2 clients]# duc index --verbose earthscience
Writing to database "/mnt/gluster/durable/rcg/duc/gstore.ducdb"
^C
[root@gstore2 clients]# duc index --verbose --uncompressed earthscience
Writing to database "/mnt/gluster/durable/rcg/duc/gstore.ducdb"
^C
[root@gstore2 clients]# duc info
[root@gstore2 clients]# time duc index --debug earthscience
Writing to database "/mnt/gluster/durable/rcg/duc/gstore.ducdb"
>> /mnt/gluster/durable/clients/earthscience
  r 12724224 12724224 .duc.db
>> dmsadmin
>> dmsadmin
>> dms
<< dms files:0 dirs:0 actual:0 apparent:0
  d 146 5632 dms
>> es_sgrlab
>> .TemporaryItems
>> folders.501
>> TemporaryItems
<< TemporaryItems files:0 dirs:0 actual:0 apparent:0
  d 130 5632 TemporaryItems
<< folders.501 files:0 dirs:1 actual:130 apparent:5632
  d 655 11264 folders.501
>> folders.1352872978
<< folders.1352872978 files:0 dirs:0 actual:0 apparent:0
  d 130 5632 folders.1352872978
>> folders.505
<< folders.505 files:0 dirs:0 actual:0 apparent:0
  d 130 5632 folders.505
^C

real    0m0.230s
user    0m0.004s
sys     0m0.004s
[root@gstore2 clients]# duc info
[root@gstore2 clients]#
```

New ticket? Is this a bug or user error?

**Comment by zevv** [comment]
* On 2015-04-25 00:26:34 +0200, pol llovet wrote:
 
> This is probably the wrong place for this, but something is awry. Should I create a new issue?

Yes, let's do that, it seems serious enough to deserve it's own ticket.

In the mean time, could you please try to:

- remove the database file and try again? In the past I have seen a
  tokyocabinet database go corrupt, without any of the API functions
  generating errors.

- store the database on a local file system instead on glusterfs. This
  should not be a problem, but maybe tokyocabinet has problems with
  the underlying file system.

Does duc work for you as it should on local filesystems like ext4?

Thanks,

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by zevv** [comment]
All implementations create considerable complexity with which I'm not willing to cope at this time. There is probably much more gain to be expected from re-using previous index results during caching, so closing this ticket for now.
